src/extensions/store/sdk.ts

```diff
@@ -0,0 +1,215 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { createStore } from "zustand/vanilla";
+import { produce } from "immer";
+import { TfheCompactPublicKey } from "../../sdk/fhe/fhe";
+import { InitFhevm } from "../../sdk/init";
+import { fromHexString, toABIEncodedUint32 } from "../../sdk/utils";
+import { FheOpsAddress, PUBLIC_KEY_LENGTH_MIN } from "../../sdk/consts";
+import { SendFn, SignTypedDataFn } from "../../sdk/types";
+
+type ChainRecord<T> = Record<string, T>;
+type SecurityZoneRecord<T> = Record<number, T>;
+
+type SdkStore = {
+  initialized: boolean;
+  fhevmInitialized: boolean;
+  fheKeysInitialized: boolean;
+
+  securityZones: number[];
+  fheKeys: ChainRecord<SecurityZoneRecord<Uint8Array | undefined>>;
+
+  account?: string;
+  chainId?: string;
+  send?: SendFn;
+  signTypedData?: SignTypedDataFn;
+};
+
+export const _sdkStore = createStore<SdkStore>(
+  () =>
+    ({
+      initialized: false,
+      fhevmInitialized: false,
+      fheKeysInitialized: false,
+
+      securityZones: [0],
+      fheKeys: {},
+
+      account: undefined,
+      chainId: undefined,
+      send: undefined,
+      signTypedData: undefined,
+    }) as SdkStore,
+);
+
+export const _store_getFheKey = (
+  chainId: string | undefined,
+  securityZone: number | undefined,
+): TfheCompactPublicKey | undefined => {
+  if (chainId == null || securityZone == null) return undefined;
+
+  const serialized = _sdkStore.getState().fheKeys[chainId]?.[securityZone];
+  if (serialized == null) return undefined;
+
+  return TfheCompactPublicKey.deserialize(serialized);
+};
+
+export const _store_getConnectedChainFheKey = (
+  securityZone: number | undefined,
+): TfheCompactPublicKey | undefined => {
+  const state = _sdkStore.getState();
+
+  if (securityZone == null) return undefined;
+  if (state.chainId == null) return undefined;
+
+  const serialized = state.fheKeys[state.chainId]?.[securityZone];
+  if (serialized == null) return undefined;
+
+  return TfheCompactPublicKey.deserialize(serialized);
+};
+
+export const _store_setFheKey = (
+  chainId: string | undefined,
+  securityZone: number | undefined,
+  fheKey: TfheCompactPublicKey | undefined,
+) => {
+  if (chainId == null || securityZone == null) return;
+
+  _sdkStore.setState(
+    produce<SdkStore>((state) => {
+      if (state.fheKeys[chainId] == null) state.fheKeys[chainId] = {};
+      state.fheKeys[chainId][securityZone] = fheKey?.serialize();
+    }),
+  );
+};
+
+const getChainIdFromProvider = async (send: SendFn): Promise<string> => {
+  const chainId = (await send("eth_chainId").catch((err: Error) => {
+    throw Error(`Error while requesting chainId from provider: ${err}`);
+  })) as string;
+
+  if (isNaN(parseInt(chainId, 16))) {
+    throw new Error(
+      `received non-hex number from chainId request: "${chainId}"`,
+    );
+  }
+
+  return chainId;
+};
+
+export type InitParams = {
+  account: string;
+  send: SendFn;
+  signTypedData: SignTypedDataFn;
```

Comment:
Is this an intuitive interface? Knowing how to pass these methods as parameters sounds a bit challenging I would think

Reply:
The intent was to allow users to be provider agnostic, and to allow viem based "providers" (referred to as clients in viem) to interact with the sdk.

I'm happy to revert this, I'm not convinced that this is better than the existing version.